### ISSUE TITLE
Bug: Openvpn stopping but not starting on restart calls

### ISSUE THREAD

**Issue by qdm12** [qdm12/gluetun#389]
**Is this urgent?**: No

**What is the version of the program** (See the line at the top of your logs)

```
Running version latest built on 2021-02-12T21:37:24Z (commit adaad62)
```

**What's the problem** 🤔

The openvpn loop gets stopped but not started on restart calls. This is likely due to a deadlock somewhere in my code.

**Share your logs...**

```log
2021-02-13T06:37:02.067Z	INFO	openvpn: [us8375.nordvpn.com] Peer Connection Initiated with [AF_INET]138.199.52.56:1194
2021-02-13T06:37:03.198Z	INFO	openvpn: TUN/TAP device tun0 opened
2021-02-13T06:37:03.198Z	INFO	openvpn: /sbin/ip link set dev tun0 up mtu 1500
2021-02-13T06:37:03.200Z	INFO	openvpn: /sbin/ip addr add dev tun0 10.8.3.2/24 broadcast 10.8.3.255
2021-02-13T06:37:03.204Z	INFO	openvpn: Initialization Sequence Completed
2021-02-13T06:41:01.388Z	INFO	http server: 200 GET /ip wrote 31B to 172.18.0.10:33130 in 16.891µs
2021-02-13T06:41:01.390Z	INFO	openvpn: stopping
2021-02-13T06:41:01.417Z	INFO	http server: openvpn: stopped
2021-02-13T06:41:01.417Z	INFO	openvpn: starting
2021-02-13T06:41:01.420Z	INFO	firewall: setting VPN connection through firewall...
2021-02-13T06:41:01.427Z	INFO	openvpn configurator: starting openvpn
2021-02-13T06:41:01.428Z	INFO	http server: openvpn: running
2021-02-13T06:41:01.428Z	INFO	http server: 200 GET /openvpn/actions/restart wrote 68B to 172.18.0.10:33132 in 37.726024ms
2021-02-13T06:41:01.430Z	INFO	openvpn: OpenVPN 2.4.10 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Jan  4 2021
2021-02-13T06:41:01.430Z	INFO	openvpn: library versions: OpenSSL 1.1.1i  8 Dec 2020, LZO 2.10
2021-02-13T06:41:01.430Z	WARN	openvpn: --ping should normally be used with --ping-restart or --ping-exit
2021-02-13T06:41:01.433Z	INFO	openvpn: TCP/UDP: Preserving recently used remote address: [AF_INET]184.170.247.169:1194
2021-02-13T06:41:01.433Z	INFO	openvpn: UDP link local: (not bound)
2021-02-13T06:41:01.433Z	INFO	openvpn: UDP link remote: [AF_INET]184.170.247.169:1194
2021-02-13T06:41:02.344Z	INFO	openvpn: [us6190.nordvpn.com] Peer Connection Initiated with [AF_INET]184.170.247.169:1194
2021-02-13T06:41:03.616Z	INFO	openvpn: TUN/TAP device tun0 opened
2021-02-13T06:41:03.616Z	INFO	openvpn: /sbin/ip link set dev tun0 up mtu 1500
2021-02-13T06:41:03.617Z	INFO	openvpn: /sbin/ip addr add dev tun0 10.8.3.2/24 broadcast 10.8.3.255
2021-02-13T06:41:03.621Z	INFO	openvpn: Initialization Sequence Completed
2021-02-13T06:45:01.531Z	INFO	http server: 200 GET /ip wrote 31B to 172.18.0.10:34440 in 21.233µs
2021-02-13T06:45:01.534Z	INFO	openvpn: stopping
2021-02-13T06:45:01.563Z	INFO	http server: openvpn: stopped
2021-02-13T06:45:01.563Z	INFO	openvpn: starting
2021-02-13T06:45:01.568Z	INFO	firewall: setting VPN connection through firewall...
2021-02-13T06:45:01.574Z	INFO	openvpn configurator: starting openvpn
2021-02-13T06:45:01.574Z	INFO	http server: openvpn: running
2021-02-13T06:45:01.574Z	INFO	http server: 200 GET /openvpn/actions/restart wrote 68B to 172.18.0.10:34442 in 40.278345ms
2021-02-13T06:45:01.577Z	INFO	openvpn: OpenVPN 2.4.10 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Jan  4 2021
2021-02-13T06:45:01.577Z	INFO	openvpn: library versions: OpenSSL 1.1.1i  8 Dec 2020, LZO 2.10
2021-02-13T06:45:01.578Z	WARN	openvpn: --ping should normally be used with --ping-restart or --ping-exit
2021-02-13T06:45:01.580Z	INFO	openvpn: TCP/UDP: Preserving recently used remote address: [AF_INET]217.138.200.59:1194
2021-02-13T06:45:01.580Z	INFO	openvpn: UDP link local: (not bound)
2021-02-13T06:45:01.580Z	INFO	openvpn: UDP link remote: [AF_INET]217.138.200.59:1194
2021-02-13T06:45:02.214Z	INFO	openvpn: [ca1379.nordvpn.com] Peer Connection Initiated with [AF_INET]217.138.200.59:1194
2021-02-13T06:45:03.561Z	INFO	openvpn: TUN/TAP device tun0 opened
2021-02-13T06:45:03.562Z	INFO	openvpn: /sbin/ip link set dev tun0 up mtu 1500
2021-02-13T06:45:03.563Z	INFO	openvpn: /sbin/ip addr add dev tun0 10.8.1.2/24 broadcast 10.8.1.255
2021-02-13T06:45:03.569Z	INFO	openvpn: Initialization Sequence Completed
2021-02-13T06:49:01.677Z	INFO	http server: 200 GET /ip wrote 31B to 172.18.0.10:35748 in 18.338µs
2021-02-13T06:49:01.683Z	INFO	openvpn: stopping
2021-02-13T07:31:50.847Z	INFO	healthcheck: unhealthy: lookup github.com on 127.0.0.1:53: read udp 127.0.0.1:48688->127.0.0.1:53: i/o timeout
2021-02-13T08:59:30.124Z	INFO	http server: 404 GET  wrote 41B to 74.120.14.55:59088 in 13.086µs
2021-02-13T08:59:30.318Z	INFO	http server: 404 GET  wrote 41B to 74.120.14.55:59614 in 11.94µs
2021-02-13T11:06:34.508Z	ERROR	ip getter: Get "https://domains.google.com/checkip": dial tcp: lookup domains.google.com on 127.0.0.1:53: server misbehaving
2021-02-13T11:06:34.508Z	INFO	ip getter: retrying in 5s
2021-02-13T11:06:39.510Z	ERROR	ip getter: Get "https://ifconfig.io/ip": dial tcp: lookup ifconfig.io on 127.0.0.1:53: server misbehaving
2021-02-13T11:06:39.510Z	INFO	ip getter: retrying in 5s
2021-02-13T11:06:44.511Z	ERROR	ip getter: Get "https://domains.google.com/checkip": dial tcp: lookup domains.google.com on 127.0.0.1:53: server misbehaving
2021-02-13T11:06:44.511Z	INFO	ip getter: retrying in 5s
```

**Comment by darktohka** [comment]
**Is this urgent?**: No

**Host OS** (approximate answer is fine too): Manjaro

**CPU arch** or **device name**: amd64

**What VPN provider are you using**: Windscribe

**What are you using to run your container?**: Docker Compose

**What is the version of the program** (See the line at the top of your logs)

```
Running version latest built on 2021-05-23T21:55:31Z (commit 0c4f01a)
```

**What's the problem** 🤔

Using Windscribe as a VPN provider. I have two VPN containers running simultaneously. Sometimes, the VPN locks up: the OpenVPN process exits, GlueTun says it's `restarting openvpn`, but the OpenVPN process never reboots. The service becomes unhealthy indefinitely and does not recover.

According to `docker-compose top`, these are the processes running in the container:
```
service_vpn_1
UID      PID      PPID     C   STIME   TTY     TIME         CMD
-------------------------------------------------------------------
root   2023733   2023706   0   May26   ?     00:00:46   /entrypoint
```
Here's an example for a healthy service that is currently running perfectly fine:
```
service_vpn2_1
UID      PID      PPID     C   STIME   TTY     TIME                        CMD
-------------------------------------------------------------------------------------------------
root   2023515   2023476   0   May26   ?     00:00:23   /entrypoint
root   2023697   2023515   0   May26   ?     00:00:04   openvpn --config /etc/openvpn/target.ovpn
```

I've decided not to create a separate issue as this issue seems similar enough.

**Share your logs... (careful to remove in example tokens)**

https://fastbin.xyz/zi00in9

**Comment by qdm12** [comment]
Thanks! Yeah definitely the same issue (I hope). It's a bit of a spaghetti asynchronous code unfortunately and the 'openvpn loop' handling the state of the openvpn subprocess seems to stuck itself occasionally. I'll try to improve the code and add tests soon. Thanks!

**Comment by qdm12** [comment]
I've been changing a few things in the code recently (only on `:latest`), and have been running gluetun with a 30 minutes restart timer for the last 12 hours with success, so far no stuck restart. I'll leave it for a few more days, but the issue might be resolved.

**Comment by qdm12** [comment]
I think it should be fixed especially since 39068dda1713ad25aa0e0ce72a4f0acf3291e48c

Please you guys try it and let me know after a few days if you encounter the issue again, thanks!

**Comment by qdm12** [comment]
Closing the issue for now, feel free to comment if it doesn't work!
